MG-2117 - Remove repository errors from API layer

WashingtonKK opened the pull request:
# What type of PR is this?

This is a feature that removes removes repository errors from the api response and only the service error and api errors are sent as part of the response. All errors are logged and can be evaluated from the service logs.

## What does this do?

Ensures only errors from the service and api are sent as part of the API response, and leaves repo errors at service level.

## Which issue(s) does this PR fix/relate to?

- Resolves #2117
- Related to #1998 

## Have you included tests for your changes?

No

## Did you document any new/modified feature?

No

### Notes

WashingtonKK committed:
feat: return only service errors

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Remove repoerr from things service

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Remove repoerr in services

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Fix ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Add servive error check

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Wrap permission list error

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Wrap permission list error

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Modify tests and keep internal errors

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Fix tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: enhance error encoding

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: update api docs

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

feat: update checkdomain and checkpolicy

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: errors on service layer

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

update tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: property based tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Remove repo error from service layer

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix test'

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

refactor errors:

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: property based tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

enhance errors

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

remove unused error

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

update bootstrap and certs tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Update docs

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

update bootstrap test

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix auth tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix typo:

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

arvindh123 commented on users/service.go:
```diff
@@ -64,10 +44,10 @@ func (svc service) RegisterClient(ctx context.Context, token string, cli mgclien
 		userID, err := svc.Identify(ctx, token)
 		if err != nil {
-			return mgclients.Client{}, err
+			return mgclients.Client{}, errors.Wrap(svcerr.ErrAuthorization, err)
```
In Line 45 `svc.Identify(ctx, token)` already wraps `svcerr.ErrAuthorization`, so need to wrap it again here 
If the functions call is goes out the package then wrapping is required, function call is within the package, the called function should wrap right  errors

arvindh123 commented on users/service.go:
```diff
@@ -64,10 +44,10 @@ func (svc service) RegisterClient(ctx context.Context, token string, cli mgclien
 		}
 		if err := svc.checkSuperAdmin(ctx, userID); err != nil {
-			return mgclients.Client{}, err
+			return mgclients.Client{}, errors.Wrap(svcerr.ErrAuthorization, err)
```
`svc.checkSuperAdmin(ctx, userID)` function wraps `svcerr.ErrAuthorization`, so need to wrap it again here 
If the functions call is goes out the package then wrapping is required, function call is within the package, the called function should wrap right  errors

arvindh123 commented on users/service.go:
```diff
@@ -94,7 +74,7 @@ func (svc service) RegisterClient(ctx context.Context, token string, cli mgclien
 
 	if err := svc.addClientPolicy(ctx, cli.ID, cli.Role); err != nil {
-		return mgclients.Client{}, err
+		return mgclients.Client{}, errors.Wrap(svcerr.ErrInvalidPolicy, err)
```
same here

arvindh123 commented on users/service.go:
```diff
@@ -517,7 +513,7 @@ func (svc service) retrieveObjectUsersPermissions(ctx context.Context, domainID,
 	permissions, err := svc.listObjectUserPermission(ctx, userID, objectType, objectID)
 	if err != nil {
-		return err
+		return errors.Wrap(svcerr.ErrFailedPermissionsList, err)
```
From user prespective , this can be a internal error.
@WashingtonKK What do you think? Please give your comments

WashingtonKK replied:
Yes, it needs to be an internal error but we still need to notify the user of an error.

We can change it to an errAuthorization instead.
What do you think about this @arvindh123 ?

arvindh123 replied:
I see  `svc.listObjectUserPermission` already wrap with error `svcerr.ErrFailedPermissionsList`, we can make `svcerr.ErrFailedPermissionsList` as internal error and return `svcerr.ErrAuthroization` or something like denied

WashingtonKK replied:
This make sense, okay

arvindh123 commented on users/service.go:
```diff
@@ -581,9 +577,9 @@ func (svc service) OAuthCallback(ctx context.Context, state mgoauth2.State, clie
 		if err != nil {
 			if errors.Contains(err, repoerr.ErrNotFound) {
-				return &magistrala.Token{}, errors.New("user not signed up")
+				return &magistrala.Token{}, errUserNotSignedUp
```
This one could not be a internal error,we can simply return  ErrAuthorization
@WashingtonKK Please give your comments

WashingtonKK replied:
Here it is better to notify that the user is not signed up so that the user can sign up.

WashingtonKK replied:
I think it is better to return errUserNotSignedUp here @arvindh123

arvindh123 replied:
This endpoint presents a potential risk of user data leakage, as unauthorized individuals could potentially scrape data by sending requests to it. If they encounter an authorization error, they can confirm whether a user's email is registered within the system. Conversely, if the endpoint returns "UserNotSignedUp," they can deduce that the user is not registered.

WashingtonKK replied:
Okay, let me wrap the error with errAuthorization so that userNotSignedUp can remain internal

arvindh123 commented on things/service.go:
```diff
@@ -125,7 +119,7 @@ func (svc service) CreateThings(ctx context.Context, token string, cls ...mgclie
 	}
 	if _, err := svc.auth.AddPolicies(ctx, &policies); err != nil {
-		return nil, errors.Wrap(errAddPolicies, err)
+		return nil, errors.Wrap(svcerr.ErrAddPolicies, err)
```
I think, this can be an internal error. The user won't doesn't need information about backend errors.

arvindh123 commented on users/service.go:
```diff
@@ -704,10 +701,10 @@ func (svc service) updateClientPolicy(ctx context.Context, userID string, role m
 		})
 		if err != nil {
-			return errors.Wrap(errDeletePolicies, err)
+			return errors.Wrap(svcerr.ErrDeletePolicies, err)
```
Same here, I think for user information about policy error might not be useful.

arvindh123 commented on auth/service_test.go:
```diff
@@ -1975,7 +1975,7 @@ func TestChangeDomainStatus(t *testing.T) {
 			},
 			retreieveByIDErr: repoerr.ErrNotFound,
-			err:              svcerr.ErrAuthorization,
+			err:              svcerr.ErrNotFound,
```
@WashingtonKK, Why we are changing this error type?

arvindh123 replied:
If the domain is empty or not avialable with MG, then it should return Authorization Error, 
We should not return 404 , So I'm suggesting following changes in two functions 


1.) 

https://github.com/washingtonkk/magistrala/blob/b50c99851ec8b3bbb23b26476896c1aad8f12fde/auth/service.go#L187-L205
Suggestion  for the above function 
```go
func (svc service) checkPolicy(ctx context.Context, pr PolicyReq) error {
	// Domain status is required for if user sent authorization request on things, channels, groups and domains
	if pr.SubjectType == UserType && (pr.ObjectType == GroupType || pr.ObjectType == ThingType || pr.ObjectType == DomainType) {
		if err := svc.checkDomain(ctx, pr.SubjectType, pr.Subject, pr.Domain); err != nil {
			return err
		}
	}
	if err := svc.agent.CheckPolicy(ctx, pr); err != nil {
		return errors.Wrap(svcerr.ErrAuthorization, err)
	}
	return nil
}
```

2.) 
https://github.com/washingtonkk/magistrala/blob/b50c99851ec8b3bbb23b26476896c1aad8f12fde/auth/service.go#L207-L240
Suggestion for above function 

```
func (svc service) checkDomain(ctx context.Context, subjectType, subject, domainID string) error {
	if err := svc.agent.CheckPolicy(ctx, PolicyReq{
		Subject:     subject,
		SubjectType: subjectType,
		Permission:  MembershipPermission,
		Object:      domainID,
		ObjectType:  DomainType,
	}); err != nil {
		return svcerr.ErrDomainAuthorization
	}

	d, err := svc.domains.RetrieveByID(ctx, domainID)
	if err != nil {
		return errors.Wrap(errors.ErrUnidentified, err)
	}

	switch d.Status {
	case EnabledStatus:
	case DisabledStatus:
		if err := svc.agent.CheckPolicy(ctx, PolicyReq{
			Subject:     subject,
			SubjectType: subjectType,
			Permission:  AdminPermission,
			Object:      domainID,
			ObjectType:  DomainType,
		}); err != nil {
			return svcerr.ErrDomainAuthorization
		}
	case FreezeStatus:
		if err := svc.agent.CheckPolicy(ctx, PolicyReq{
			Subject:     subject,
			SubjectType: subjectType,
			Permission:  AdminPermission,
			Object:      MagistralaObject,
			ObjectType:  PlatformType,
		}); err != nil {
			return svcerr.ErrDomainAuthorization
		}
	default:
		return svcerr.ErrDomainAuthorization
	}

	return nil
}
```

WashingtonKK replied:
This makes sense @arvindh123, it is a great idea. Let me implement it

arvindh123 replied:
Thank for considering suggestions @WashingtonKK

rodneyosodo commented on things/service.go:
```diff
@@ -125,7 +119,7 @@ func (svc service) CreateThings(ctx context.Context, token string, cls ...mgclie
 	}
 	if _, err := svc.auth.AddPolicies(ctx, &policies); err != nil {
-		return nil, errors.Wrap(errAddPolicies, err)
+		return nil, errors.Wrap(svcerr.ErrCreateEntity, err)
```
Also add `svcerr.ErrAddPolicies`

WashingtonKK replied:
We do not want ```svcerr.ErrAddPolicies``` to be sent to a user, that is why we send ```svcerr.ErrCreateEntity``` which is what the user will see. errAddPolicies is left as an internal error which will be logged but we do not want it to be sent along with the response

rodneyosodo commented on things/service.go:
```diff
@@ -411,10 +405,10 @@ func (svc service) Share(ctx context.Context, token, id, relation string, userid
 	res, err := svc.auth.AddPolicies(ctx, &policies)
 	if err != nil {
-		return errors.Wrap(errAddPolicies, err)
+		return errors.Wrap(svcerr.ErrUpdateEntity, err)
```
same

WashingtonKK replied:
We do not want ```svcerr.ErrAddPolicies``` to be sent to a user, that is why we send ```svcerr.ErrCreateEntity``` which is what the user will see ```errAddPolicies``` is left as an internal error which will be logged but we do not want it to be sent along with the response

rodneyosodo commented on things/service.go:
```diff
@@ -411,10 +405,10 @@ func (svc service) Share(ctx context.Context, token, id, relation string, userid
 	}
 	if !res.Added {
-		return err
+		return errors.Wrap(svcerr.ErrUpdateEntity, err)
```
Don't wrap the error since it has already been handled above just return `svcerr.ErrAddPolicies`

WashingtonKK replied:
We do not want ```svcerr.ErrAddPolicies``` to be sent to a user, that is why we send ```svcerr.ErrCreateEntity``` which is what the user will see ```errAddPolicies``` is left as an internal error which will be logged but we do not want it to be sent along with the response

rodneyosodo commented on things/service.go:
```diff
@@ -440,7 +434,7 @@ func (svc service) Unshare(ctx context.Context, token, id, relation string, user
 	res, err := svc.auth.DeletePolicies(ctx, &policies)
 	if err != nil {
-		return errors.Wrap(errRemovePolicies, err)
+		return errors.Wrap(svcerr.ErrUpdateEntity, err)
```
use `svcerr.ErrDeletePolicies` instead

WashingtonKK replied:
We do not want ```svcerr.ErrDeletePolicies``` to be sent to a user, that is why we send ```svcerr.ErrUpdateEntity``` which is what the user will see.

rodneyosodo commented on things/service_test.go:
```diff
@@ -43,8 +43,7 @@ var (
 	wrongID           = testsutil.GenerateUUID(&testing.T{})
-	errAddPolicies    = errors.New("failed to add policies")
-	errRemovePolicies = errors.New("failed to remove the policies")
+	errRemovePolicies = errors.New("failed to delete policies")
```
no need for this use errors from `svcerr`

WashingtonKK replied:
We want these to be internal errors, as mentioned above. So we define them within the service so that they can be logged. Errors in the svcerr package are what we want to be added to the request body.

rodneyosodo commented on users/service.go:
```diff
@@ -19,6 +19,25 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
+var (
```
No need for this errors use errors from `svcerr`

WashingtonKK replied:
We need them locally and not on svcerr as described in the comments above.

rodneyosodo commented on users/service.go:
```diff
@@ -19,6 +19,25 @@ import (
+	ErrRecoveryToken = errors.New("failed to generate password recovery token")
+
+	// ErrFailedPolicyUpdate indicates a failure to update user policy.
+	ErrFailedPolicyUpdate = errors.New("failed to update user policy")
```
No need to export them if they are not used outside the package.

rodneyosodo commented on users/service.go:
```diff
@@ -288,7 +307,7 @@ func (svc service) ResetSecret(ctx context.Context, resetToken, secret string) e
 	secret, err = svc.hasher.Hash(secret)
 	if err != nil {
-		return err
+		return errors.Wrap(svcerr.ErrAuthorization, err)
```
This is not an authorization error

rodneyosodo commented on users/service.go:
```diff
@@ -399,10 +418,10 @@ func (svc service) DisableClient(ctx context.Context, token, id string) (mgclien
 	tokenUserID, err := svc.Identify(ctx, token)
 	if err != nil {
-		return mgclients.Client{}, err
+		return mgclients.Client{}, errors.Wrap(svcerr.ErrAuthorization, err)
```
```suggestion
		return mgclients.Client{}, errors.Wrap(svcerr.ErrAuthentication, err)
```

arvindh123 replied:
@rodneyosodo @WashingtonKK No need to wrap here , because it is already wrapped in `svc.Identify` , which is present in the same layer.

rodneyosodo commented on users/service.go:
```diff
@@ -399,10 +418,10 @@ func (svc service) DisableClient(ctx context.Context, token, id string) (mgclien
 	}
 	if err := svc.checkSuperAdmin(ctx, tokenUserID); err != nil {
-		return mgclients.Client{}, err
+		return mgclients.Client{}, errors.Wrap(svcerr.ErrAuthentication, err)
```
```suggestion
		return mgclients.Client{}, errors.Wrap(svcerr.ErrAuthorization, err)
```

rodneyosodo commented on users/service.go:
```diff
@@ -562,9 +581,9 @@ func (svc service) OAuthCallback(ctx context.Context, state mgoauth2.State, clie
 		if err != nil {
 			if errors.Contains(err, repoerr.ErrNotFound) {
-				return &magistrala.Token{}, errors.New("user not signed up")
+				return &magistrala.Token{}, errors.Wrap(svcerr.ErrAuthorization, errUserNotSignedUp)
```
This is not an authorization error

rodneyosodo commented on users/service.go:
```diff
@@ -562,9 +581,9 @@ func (svc service) OAuthCallback(ctx context.Context, state mgoauth2.State, clie
+				return &magistrala.Token{}, errors.Wrap(svcerr.ErrAuthorization, errUserNotSignedUp)
 			}
-			return &magistrala.Token{}, err
+			return &magistrala.Token{}, errors.Wrap(svcerr.ErrAuthorization, err)
```
same

rodneyosodo commented on users/service.go:
```diff
@@ -575,9 +594,9 @@ func (svc service) OAuthCallback(ctx context.Context, state mgoauth2.State, clie
 		if err != nil {
 			if errors.Contains(err, repoerr.ErrConflict) {
-				return &magistrala.Token{}, errors.New("user already exists")
+				return &magistrala.Token{}, errors.Wrap(svcerr.ErrConflict, err)
```
same

rodneyosodo commented on users/service.go:
```diff
@@ -376,7 +395,7 @@ func (svc service) EnableClient(ctx context.Context, token, id string) (mgclient
 	client, err := svc.changeClientStatus(ctx, token, client)
 	if err != nil {
-		return mgclients.Client{}, errors.Wrap(mgclients.ErrEnableClient, err)
+		return mgclients.Client{}, err
```
revert to give more context to the error

rodneyosodo commented on certs/service.go:
```diff
@@ -154,7 +153,7 @@ func (cs *certsService) ListCerts(ctx context.Context, token, thingID string, of
 	cp, err := cs.certsRepo.RetrieveByThing(ctx, u.GetId(), thingID, offset, limit)
 	if err != nil {
-		return Page{}, errors.Wrap(repoerr.ErrNotFound, err)
+		return Page{}, errors.Wrap(svcerr.ErrNotFound, err)
```
Should this be an `ErrViewEntity` or a `ErrNotFound`

arvindh123 commented on users/service.go:
```diff
@@ -260,15 +274,15 @@ func (svc service) UpdateClientIdentity(ctx context.Context, token, clientID, id
 	client, err := svc.clients.RetrieveByIdentity(ctx, email)
 	if err != nil || client.Credentials.Identity == "" {
-		return repoerr.ErrNotFound
+		return svcerr.ErrNotFound
```
This can lead to data leak, Any one can check user exists in platform or not, by requesting reset token   
@dborovcanin & @rodneyosodo  What do you think about this approach

rodneyosodo replied:
Refactor this to something like
```go
client, err := svc.clients.RetrieveByIdentity(ctx, email)
if err != nil {
	return errors.Wrap(svcerr.ErrViewEntity, err)
}
```

This also `ResetSecret` where we check that identity is not empty

arvindh123 commented on api/openapi/bootstrap.yml:
```diff
@@ -96,6 +96,8 @@ paths:
           $ref: "#/components/responses/ConfigRes"
+        "400":
+          $description: Missing or invalid config.
         "401":
```
```suggestion
          description: Missing or invalid config.
```

arvindh123 reviewed (changes requested):
https://github.com/absmach/magistrala/pull/2119#discussion_r1549247412

Second part is not addressed

WashingtonKK committed:
update auth api docs

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

refactor errors

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

refactor errors

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

rodneyosodo commented on users/service.go:
```diff
@@ -284,11 +298,11 @@ func (svc service) ResetSecret(ctx context.Context, resetToken, secret string) e
 	}
 	if c.Credentials.Identity == "" {
-		return repoerr.ErrNotFound
+		return svcerr.ErrNotFound
```
Remove this check no need to check `c.Credentials.Identity` is empty as it is always present

rodneyosodo commented on users/service.go:
```diff
@@ -575,9 +589,9 @@ func (svc service) OAuthCallback(ctx context.Context, state mgoauth2.State, clie
 		if err != nil {
 			if errors.Contains(err, repoerr.ErrConflict) {
-				return &magistrala.Token{}, errors.New("user already exists")
+				return &magistrala.Token{}, errors.Wrap(svcerr.ErrConflict, err)
```
I don't think we need to wrap with `svcerr.ErrConflict` since it contains `repoerr.ErrConflict` and they have the same error message

WashingtonKK replied:
I don't understand, `repoerr.ErrConflict` has message `entity already exists`.

rodneyosodo replied:
`svcerr.ErrConflict` and `repoerr.ErrConflict` have the same error message the difference is one should be used at service layer while the other should be used in repository

rodneyosodo commented on users/service.go:
```diff
@@ -19,6 +19,20 @@ import (
+var (
+
+	// ErrFailedUpdateRole indicates a failure to update user role.
+	ErrFailedUpdateRole = errors.New("failed to update user role")
```
No need to export this if it is not being used outside this package

rodneyosodo commented on users/service.go:
```diff
@@ -19,6 +19,20 @@ import (
+	ErrFailedUpdateRole = errors.New("failed to update user role")
+
+	// ErrIssueToken indicates a failure to issue token.
+	ErrIssueToken = errors.New("failed to issue token")
```
same

rodneyosodo commented on pkg/errors/service/types.go:
```diff
@@ -55,18 +55,23 @@ var (
 	ErrFailedUpdateRole = errors.New("failed to update user role")
+
+	// ErrFailedPermissionsList indicates a failure to list permissions.
+	ErrFailedPermissionsList = errors.New("failed to list permissions")
```
Add new line after

rodneyosodo commented on certs/service.go:
```diff
@@ -186,7 +185,7 @@ func (cs *certsService) ViewCert(ctx context.Context, token, serialID string) (C
 	cert, err := cs.certsRepo.RetrieveBySerial(ctx, u.GetId(), serialID)
 	if err != nil {
-		return Cert{}, errors.Wrap(repoerr.ErrNotFound, err)
+		return Cert{}, errors.Wrap(svcerr.ErrNotFound, err)
```
This should be. Also check other cases where we wrap `ErrNotFound` when we fail to view an entity from the DB.
```suggestion
		return Cert{}, errors.Wrap(svcerr.ErrViewEntity, err)
```

rodneyosodo commented on users/service.go:
```diff
@@ -283,12 +291,10 @@ func (svc service) ResetSecret(ctx context.Context, resetToken, secret string) e
 	secret, err = svc.hasher.Hash(secret)
 	if err != nil {
-		return err
+		return errors.Wrap(svcerr.ErrUpdateEntity, err)
```
```suggestion
		return errors.Wrap(svcerr.ErrMalformedEntity, err)
```

arvindh123 commented on users/service.go:
```diff
@@ -259,16 +267,16 @@ func (svc service) UpdateClientIdentity(ctx context.Context, token, clientID, id
 	client, err := svc.clients.RetrieveByIdentity(ctx, email)
-	if err != nil || client.Credentials.Identity == "" {
-		return repoerr.ErrNotFound
+	if err != nil {
+		return errors.Wrap(svcerr.ErrViewEntity, err)
 	}
```
@rodneyosodo  @WashingtonKK @dborovcanin 

What do you think about this suggestion ? 
```suggestion
	if err != nil {
		if errors.Contains(repoerr.ErrNotFound, err) {
			return nil
		} else {
			return errors.Wrap(svcerr.ErrViewEntity, err) // it should be an internal error
		}
	}
```

WashingtonKK replied:
Why would we return nil when the error contains `ErrNotFound` ? Shouldn't we return `svcerr.ErrNotFound` in this case?

WashingtonKK committed:
fix: certs and twins tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: certs and twins tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

arvindh123 commented on users/service.go:
```diff
@@ -399,10 +405,10 @@ func (svc service) DisableClient(ctx context.Context, token, id string) (mgclien
 	}
 	if err := svc.checkSuperAdmin(ctx, tokenUserID); err != nil {
-		return mgclients.Client{}, err
+		return mgclients.Client{}, errors.Wrap(svcerr.ErrAuthorization, err)
```
@WashingtonKK 
`svc.Identify` already returns the `svcerr.ErrAuthentication` and `svc.checkSuperAdmin`, So no need to wrap here again. 
We are calling here function which belongs same service layer.

arvindh123 commented on users/service.go:
```diff
@@ -577,7 +583,7 @@ func (svc service) OAuthCallback(ctx context.Context, state mgoauth2.State, clie
 			if errors.Contains(err, repoerr.ErrConflict) {
 				return &magistrala.Token{}, errors.New("user already exists")
```
Wrapping  this error with `svcerr.ErrConflict` will return conflict status code in HTTP response

arvindh123 commented on users/service.go:
```diff
@@ -315,11 +321,11 @@ func (svc service) UpdateClientSecret(ctx context.Context, token, oldSecret, new
 	}
 	if _, err := svc.IssueToken(ctx, dbClient.Credentials.Identity, oldSecret, ""); err != nil {
-		return mgclients.Client{}, errors.Wrap(svcerr.ErrIssueToken, err)
+		return mgclients.Client{}, errors.Wrap(errIssueToken, err)
```
In `svc.IssueToken` gRPC calls error is not handled. 
Instead of wrapping error here, we can wrap and return correct error from `svc.IssueToken`, since `svc.IssueToken` is present in same service layer.

arvindh123 commented on users/service.go:
```diff
@@ -577,7 +583,7 @@ func (svc service) OAuthCallback(ctx context.Context, state mgoauth2.State, clie
 				return &magistrala.Token{}, errors.New("user already exists")
 			}
-			return &magistrala.Token{}, err
+			return &magistrala.Token{}, errors.Wrap(svcerr.ErrCreateEntity, err)
```
This wrap can be remove since `svc.RegisterClient` is present in same user service layer.

arvindh123 reviewed (commented):
For the below request,  I get error `{"error":"","message":"view entity failed"}`

```bash
curl --location 'http://localhost/users/tokens/issue' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--data '{
        "identity": "some-random-string",
        "secret": "12345678"
}'
```

If we try to update tag of non existance user id , 
Example
```bash
curl --location --request PATCH 'http://localhost/users/unknown-user-id/tags' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'Authorization: Bearer <Token>' \
--data '{
  "tags": [
    "tag1", "tag2"
  ]
}'
```

we are getting  `{   "error": "",    "message": "update entity failed" }` in response 

Same kind of result get for below request too.
curl --location --request PATCH 'http://localhost/users/03cb42ba-4592-4f53-97c9-a776ee31a06a11/role'

WashingtonKK committed:
refactor errors in users service layer

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: issue token

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

refactor error handling and fix tests:

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix tests and remove redundant error

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Update users/service.go

Co-authored-by: Arvindh <30824765+arvindh123@users.noreply.github.com>
Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

modify delete policies error

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

revert error types

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

revert policy rollback error

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

revert domain check

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

remove debug logs

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix bootstrap property based tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

refactor bootstrap service error

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

arvindh123 reviewed (changes requested):
Issue token doesn't return proper error 

**What were you trying to achieve?**

Login to Magistrala with unknow / invalid user identity

```bash
curl --location 'http://localhost/users/tokens/issue' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--data '{
        "identity": "unknown-users-indentity",
        "secret": "12345678"
}'
```
**What are the expected results?**

HTTP/1.1 401 Unauthorized
```json
{
    "error": "",
    "message": "invalid user id or secret"
}
```

**What are the received results?**
HTTP/1.1 400 Bad Request

```json
{
    "error": "",
    "message": "entity not found"
}
```

arvindh123 commented on users/service.go:
```diff
@@ -85,15 +93,15 @@ func (svc service) RegisterClient(ctx context.Context, token string, cli mgclien
 	dbUser, err := svc.clients.RetrieveByIdentity(ctx, identity)
 	if err != nil {
-		return &magistrala.Token{}, errors.Wrap(svcerr.ErrViewEntity, err)
+		return &magistrala.Token{}, err
```
Please Wrap this error with `svcerr.ErrAuthentication`, most of the times, error here  will be repoerr.ErrNotFound, So i suggest to wrap with `svcerr.ErrAuthentication`.

arvindh123 commented on users/service.go:
```diff
@@ -619,7 +631,7 @@ func (svc service) addClientPolicy(ctx context.Context, userID string, role mgcl
 	resp, err := svc.auth.AddPolicies(ctx, &policies)
 	if err != nil {
-		return errors.Wrap(svcerr.ErrAddPolicies, err)
+		return errors.Wrap(errAddPolicies, err)
```
Can we have some thing similar to what we have in remove policy 
```suggestion
		return errors.Wrap(svcerr.ErrAddPolicies, err)
```

arvindh123 commented on users/service.go:
```diff
@@ -685,10 +697,10 @@ func (svc service) updateClientPolicy(ctx context.Context, userID string, role m
-			return errors.Wrap(svcerr.ErrDeletePolicies, err)
+			return errors.Wrap(svcerr.ErrUpdateEntity, err)
 		}
 		if !resp.Deleted {
```
Please revert back them to original, We can handle errors in ticket https://github.com/absmach/magistrala/issues/2202

arvindh123 commented on auth/service.go:
```diff
@@ -186,14 +186,7 @@ func (svc service) Authorize(ctx context.Context, pr PolicyReq) error {
-			domainID = pr.Object
-		}
-		if err := svc.checkDomain(ctx, pr.SubjectType, pr.Subject, domainID); err != nil {
+		if err := svc.checkDomain(ctx, pr.SubjectType, pr.Subject, pr.Domain); err != nil {
```
@WashingtonKK , 
Please revert back to original, My suggestion doesn't works properly for domain related operations 
It fails at domain authroization

arvindh123 reviewed (changes requested):
@WashingtonKK , 
Please revert back to original, My suggestion doesn't works properly for domain related operations 
It fails at domain authroization

arvindh123 commented on auth/service.go:
```diff
@@ -204,6 +204,23 @@ func (svc service) checkPolicy(ctx context.Context, pr PolicyReq) error {
+		Permission:  MembershipPermission,
+		Object:      domainID,
+		ObjectType:  DomainType,
+	})
```
Please remove this debug lines
```suggestion
```

dborovcanin commented on auth/service_test.go:
```diff
@@ -231,9 +248,15 @@ func TestIssue(t *testing.T) {
 				ObjectType:  auth.DomainType,
 				Permission:  auth.MembershipPermission,
 			},
+			checkPolicyRequest2: auth.PolicyReq{
```
Rename `checkPolicyRequest` fields to add more context. For example `checkPolicyReq` and `checkPolicyWithStatusReq`.

arvindh123 commented on internal/api/common.go:
```diff
@@ -70,6 +70,8 @@ const (
 	DescDir      = "desc"
 )
 
+var errAddBootstrap = errors.New("failed to add bootstrap configuration")
```
@WashingtonKK 
Instead of using `errAddBootstrap`in bootstrap,  can we use `svcerr.CreateEntity` ?

WashingtonKK replied:
Okay @arvindh123 . Although I think `errAddBootstrap` is a better option since `ErrCreateEntity` does not fully indicate the nature of the error and may be confusing to the user. What do you think about this?

arvindh123 replied:
It looks odd for me, The only un exported variable in that file.  And we are duplicating the same variable in two files

WashingtonKK replied:
The error was un exported in the bootstrap service, let me place it in `svcerr` package so that we will not have this odd looking error.

WashingtonKK committed:
fix ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
refactor test cases

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix: ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

arvindh123 reviewed (dismissed):
LGTM

rodneyosodo commented on users/service.go:
```diff
@@ -562,9 +573,9 @@ func (svc service) OAuthCallback(ctx context.Context, state mgoauth2.State, clie
+				return &magistrala.Token{}, errors.Wrap(svcerr.ErrNotFound, errUserNotSignedUp)
 			}
-			return &magistrala.Token{}, err
+			return &magistrala.Token{}, errors.Wrap(svcerr.ErrNotFound, err)
```
This should be wrapped with ViewEntity
```suggestion
			return &magistrala.Token{}, errors.Wrap(svcerr.ErrViewEntity, err)
```

WashingtonKK committed:
update oauth error

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
